### ISSUE TITLE
GitHub Exception when trying to run the PR-Agent from source

### ISSUE THREAD

**Issue by zmeir** [Codium-ai/pr-agent#30]
Hi,

I'm trying to run the PR Agent from source with [method 2](https://github.com/Codium-ai/pr-agent#method-2-run-from-source) and I'm running into the following error:
```bash
$ PYTHOPATH=. python pr_agent/cli.py --pr_url https://github.com/<MY_ORG>/<MY_REPO>/pull/134
Reviewing PR: https://github.com/<MY_ORG>/<MY_REPO>/pull/134
Traceback (most recent call last):
  File "/Users/zmeir/git/pr-agent/pr_agent/cli.py", line 27, in <module>
    run()
  File "/Users/zmeir/git/pr-agent/pr_agent/cli.py", line 22, in run
    reviewer = PRReviewer(args.pr_url, installation_id=None, cli_mode=True)
  File "/Users/zmeir/git/pr-agent/pr_agent/tools/pr_reviewer.py", line 19, in __init__
    self.git_provider = get_git_provider()(pr_url, installation_id)
  File "/Users/zmeir/git/pr-agent/pr_agent/git_providers/github_provider.py", line 29, in __init__
    self.set_pr(pr_url)
  File "/Users/zmeir/git/pr-agent/pr_agent/git_providers/github_provider.py", line 33, in set_pr
    self.pr = self._get_pr()
  File "/Users/zmeir/git/pr-agent/pr_agent/git_providers/github_provider.py", line 189, in _get_pr
    return self._get_repo().get_pull(self.pr_num)
  File "/Users/zmeir/git/pr-agent/pr_agent/git_providers/github_provider.py", line 186, in _get_repo
    return self.github_client.get_repo(self.repo)
  File "/Users/zmeir/git/pr-agent/venv/lib/python3.9/site-packages/github/MainClass.py", line 363, in get_repo
    headers, data = self.__requester.requestJsonAndCheck("GET", url)
  File "/Users/zmeir/git/pr-agent/venv/lib/python3.9/site-packages/github/Requester.py", line 442, in requestJsonAndCheck
    return self.__check(
  File "/Users/zmeir/git/pr-agent/venv/lib/python3.9/site-packages/github/Requester.py", line 487, in __check
    raise self.__createException(status, responseHeaders, data)
github.GithubException.BadCredentialsException: 401 {"message": "Bad credentials", "documentation_url": "https://docs.github.com/rest"}
```

I set my `user_token` in `pr_agent/settings/.secrets.toml` and made sure the token is authorized for MY_ORG and has the proper repo scope:  
![image](https://github.com/Codium-ai/pr-agent/assets/33152084/c7d39468-00bb-4bf1-93cb-c80cb37a82ba)  
<img width="406" alt="image" src="https://github.com/Codium-ai/pr-agent/assets/33152084/9a145e2c-5329-4d4a-af72-997aa60b28ec">  

I tested the same token with `curl` and it seems to work - for example:  
```bash
$ curl -X GET -H "Authorization:token <MY_TOEKN>" -H "Content-Type:application/json" 'https://api.github.com/repos/<MY_ORG>/<MY_REPO>/pulls?head=<MY_ORG>:<MY_BRANCH>&base=master&state=open'
[
  {
    "url": "https://api.github.com/repos/<MY_ORG>/<MY_REPO>/pulls/134",
    "id": <MY_ID>,
    "node_id": "<MY_NODE_ID>",
    "html_url": "https://github.com/<MY_ORG>/<MY_REPO>/pull/134",
    "diff_url": "https://github.com/<MY_ORG>/<MY_REPO>/pull/134.diff",
    "patch_url": "https://github.com/<MY_ORG>/<MY_REPO>/pull/134.patch",
    "issue_url": "https://api.github.com/repos/<MY_ORG>/<MY_REPO>/issues/134",
    "number": 134,
    "state": "open",
    ...
  }
]
```
```bash
$ curl -X GET -H "Authorization:token <MY_TOKEN>" -H "Content-Type:application/json" 'https://api.github.com/repos/<MY_ORG>/<MY_REPO>'
{
  "id": <MY_ID>,
  "node_id": "<MY_NODE_ID>",
  "name": "<MY_REPO>",
  "full_name": "<MY_ORG>/<MY_REPO>",
  "private": true,
  "owner": { ... }
}
```

Some technical info:
* OS: macOS Ventura 13.4.1 (22F82)
* Python: 3.9.13
* PyGithub: 1.58.2 (also tried upgrading to 1.59.0 in hopes that it will fix the issue - it didn't)

**Comment by hussam789** [comment]
@zmeir can you retry with a token that has an expiration date ( Don't use `No Expiration` ) ?

**Comment by zmeir** [comment]
Hi @hussam789, tried it again with a token with 30 days expiration (and full repo scope as before) and it still doesn't work. I also tried giving the token access to all scopes just to make sure and that didn't work either.

**Comment by hussam789** [comment]
@zmeir just the `repo` scope is enough, did you check also on a public Github repo, or you checked only on private repos?

**Comment by zmeir** [comment]
@hussam789 yes, tried it just now with https://github.com/Codium-ai/pr-agent/pull/31 and it still gives me the same error (and verified again with `curl` to see that I can still access the GitHub API with the token on this repo)

**Comment by okotek** [comment]
@zmeir Can you add 
```
from github import enable_console_debug_logging
enable_console_debug_logging()
```

Just before run() in cli.py and run again? 
What's the first request that fails? Can you run it with curl?

**Comment by zmeir** [comment]
@okotek Sure. Here's the full output:  
```
$ PYTHOPATH=. python pr_agent/cli.py --pr_url https://github.com/Codium-ai/pr-agent/pull/31
Reviewing PR: https://github.com/Codium-ai/pr-agent/pull/31
GET https://api.github.com/repos/Codium-ai/pr-agent {'Authorization': 'token (oauth token removed)', 'User-Agent': 'PyGithub/Python'} None ==> 401 {'server': 'GitHub.com', 'date': 'Wed, 12 Jul 2023 10:58:08 GMT', 'content-type': 'application/json; charset=utf-8', 'content-length': '80', 'x-github-media-type': 'github.v3; format=json', 'x-ratelimit-limit': '60', 'x-ratelimit-remaining': '52', 'x-ratelimit-reset': '1689162892', 'x-ratelimit-used': '8', 'x-ratelimit-resource': 'core', 'access-control-expose-headers': 'ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset', 'access-control-allow-origin': '*', 'strict-transport-security': 'max-age=31536000; includeSubdomains; preload', 'x-frame-options': 'deny', 'x-content-type-options': 'nosniff', 'x-xss-protection': '0', 'referrer-policy': 'origin-when-cross-origin, strict-origin-when-cross-origin', 'content-security-policy': "default-src 'none'", 'vary': 'Accept-Encoding, Accept, X-Requested-With', 'x-github-request-id': '9A68:BB04:7F2A590:80485E9:64AE8740'} {"message":"Bad credentials","documentation_url":"https://docs.github.com/rest"}
DEBUG:github.Requester:GET https://api.github.com/repos/Codium-ai/pr-agent {'Authorization': 'token (oauth token removed)', 'User-Agent': 'PyGithub/Python'} None ==> 401 {'server': 'GitHub.com', 'date': 'Wed, 12 Jul 2023 10:58:08 GMT', 'content-type': 'application/json; charset=utf-8', 'content-length': '80', 'x-github-media-type': 'github.v3; format=json', 'x-ratelimit-limit': '60', 'x-ratelimit-remaining': '52', 'x-ratelimit-reset': '1689162892', 'x-ratelimit-used': '8', 'x-ratelimit-resource': 'core', 'access-control-expose-headers': 'ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset', 'access-control-allow-origin': '*', 'strict-transport-security': 'max-age=31536000; includeSubdomains; preload', 'x-frame-options': 'deny', 'x-content-type-options': 'nosniff', 'x-xss-protection': '0', 'referrer-policy': 'origin-when-cross-origin, strict-origin-when-cross-origin', 'content-security-policy': "default-src 'none'", 'vary': 'Accept-Encoding, Accept, X-Requested-With', 'x-github-request-id': '9A68:BB04:7F2A590:80485E9:64AE8740'} {"message":"Bad credentials","documentation_url":"https://docs.github.com/rest"}
Traceback (most recent call last):
  File "/Users/zmeir/git/pr-agent/pr_agent/cli.py", line 30, in <module>
    run()
  File "/Users/zmeir/git/pr-agent/pr_agent/cli.py", line 22, in run
    reviewer = PRReviewer(args.pr_url, cli_mode=True)
  File "/Users/zmeir/git/pr-agent/pr_agent/tools/pr_reviewer.py", line 19, in __init__
    self.git_provider = get_git_provider()(pr_url)
  File "/Users/zmeir/git/pr-agent/pr_agent/git_providers/github_provider.py", line 22, in __init__
    self.set_pr(pr_url)
  File "/Users/zmeir/git/pr-agent/pr_agent/git_providers/github_provider.py", line 26, in set_pr
    self.pr = self._get_pr()
  File "/Users/zmeir/git/pr-agent/pr_agent/git_providers/github_provider.py", line 147, in _get_pr
    return self._get_repo().get_pull(self.pr_num)
  File "/Users/zmeir/git/pr-agent/pr_agent/git_providers/github_provider.py", line 144, in _get_repo
    return self.github_client.get_repo(self.repo)
  File "/Users/zmeir/git/pr-agent/venv/lib/python3.9/site-packages/github/MainClass.py", line 321, in get_repo
    headers, data = self.__requester.requestJsonAndCheck("GET", url)
  File "/Users/zmeir/git/pr-agent/venv/lib/python3.9/site-packages/github/Requester.py", line 398, in requestJsonAndCheck
    return self.__check(
  File "/Users/zmeir/git/pr-agent/venv/lib/python3.9/site-packages/github/Requester.py", line 423, in __check
    raise self.__createException(status, responseHeaders, output)
github.GithubException.BadCredentialsException: 401 {"message": "Bad credentials", "documentation_url": "https://docs.github.com/rest"}
```

And with `curl`:  
```
$ curl -X GET -H "Authorization:token <MY_TOKEN>" -H "User-Agent:PyGithub/Python" 'https://api.github.com/repos/Codium-ai/pr-agent'
{"id":662766482,"node_id":"R_kgDOJ4EDkg","name":"pr-agent","full_name":"Codium-ai/pr-agent","private":false,...}
```

**Comment by okotek** [comment]
@zmeir next I suspect the requests library. Do you mind running the following in the same Python interpreter where your run fails, replacing <token> with your token?

```
import requests

headers = {
    'Authorization': 'token <token>',
    'User-Agent': 'PyGithub/Python'
}
URL = 'https://api.github.com/repos/Codium-ai/pr-agent'
response = requests.get(URL, headers=headers)
print(response.status_code)
```

**Comment by zmeir** [comment]
@okotek yep that seems to be the problem :(
```python
import requests
headers = {
    'Authorization': 'token <MY_TOKEN>',
    'User-Agent': 'PyGithub/Python'
}
URL = 'https://api.github.com/repos/Codium-ai/pr-agent'
response = requests.get(URL, headers=headers)
print(response.status_code)
# 401
print(response.text)
# {"message":"Bad credentials","documentation_url":"https://docs.github.com/rest"}
```

I've installed the project in a fresh venv and only installed the `requirements.txt`. I'm using `requests==2.31.0` which appears to be the latest version.  
```
$ pip3 show requests
Name: requests
Version: 2.31.0
Summary: Python HTTP for Humans.
Home-page: https://requests.readthedocs.io
Author: Kenneth Reitz
Author-email: me@kennethreitz.org
License: Apache 2.0
Location: /Users/zmeir/git/pr-agent/venv/lib/python3.9/site-packages
Requires: certifi, charset-normalizer, idna, urllib3
Required-by: openai, PyGithub, python-gitlab, requests-toolbelt, tiktoken
```

Any idea what could cause this?

**Comment by hussam789** [comment]
Here are related issues i found (not sure if it will solve it):
https://github.com/PyGithub/PyGithub/issues/2431
uninstall PyGithub and then install:
```
pip install git+https://github.com/enricomi/pygithub.git@requester-auth-rework
```

**Comment by hussam789** [comment]
@zmeir Cool, So now can you run the pr-agent cli using your github token?
another comment people wrote is to delete or update `~/.netrc` file :
https://github.com/PyGithub/PyGithub/issues/1753#issuecomment-1367453823

**Comment by zmeir** [comment]
@hussam789 Yes! Thanks for all the help!

I didn't try tinkering with `~/.netrc`, I didn't want to create other issues somewhere else, and the first approach already worked. However, I did previously try to use the same token I have in `~/.netrc` and that didn't work.

Just to wrap things up, I also ran into another issue that has to do with the fact that I'm using Azure OpenAI behind a proxy server. Here are the solutions I used for both issues:

1. `github.GithubException.BadCredentialsException: 401 {"message": "Bad credentials", "documentation_url": "https://docs.github.com/rest"}`
   Taken from: https://github.com/PyGithub/PyGithub/issues/1753#issuecomment-1212987029
   Added the following lines after `self.header = header` in [github/Requester.py#L122](https://github.com/PyGithub/PyGithub/blob/001970d4a828017f704f6744a5775b4207a6523c/github/Requester.py#L122)
   ```python
   for header in self.headers:
       if str(self.headers[header]).startswith('token'):
           self.session.auth=(tuple(self.headers[header].split(' ')))
   ```
2. `aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host <MY_HOST>.openai.azure.com:443 ssl:default [Connect call failed ('<MY_HOST_IP>', 443)]`
   Taken from: https://stackoverflow.com/a/63364551/6877213
   Changed the default `trust_env` to `False` in [aiohttp/client.py#L216](https://github.com/aio-libs/aiohttp/blob/7911f1e9542afd0623b61f75c5e367c3c0e185bd/aiohttp/client.py#L216)
